src/spi.rs

```diff
@@ -10,6 +10,12 @@ use nb;
 ///
 /// - Due to how full duplex SPI works each `read` call must be preceded by a `send` call.
 ///
+/// - `read` calls only return the data received with the last `send` call.
+/// Previously received data is discarded
```

Comment:
I hope it isn't true for SPI controllers with FIFO. In case you have no overflows, you can extract data from FIFO in semantically correct order.

Reply:
You can either force device-independent implementations to read directly after every write (the rest is undefined behaviour, since it's unclear which byte gets returned [or an error]) or force the read function to only return the latest byte. Both of these are non-ideal and would have to be documented.

The first option gets worse with shared spi devices, since some device-specific implementation could rely on the rx fifo, which would break drivers that rely on the calls matching (the latest `read` call returning the data transmitted with the `write` call). It makes utilizing the tx fifo impossible for device-independent code. 

Utilizing the rx-fifo requires in-depth knowledge of your implementation, since it possibly only makes use of the spi fifos or does it's own, higher level interrupt/dma-driven stuff. If you need the performance, it may be a better idea to just bypass the hal, since it needs to be device-specific anyways.

(Dropping this constraint also makes the second one impossible)

Reply:
> You can either force device-independent implementations to read directly after every write (the rest is undefined behaviour, since it's unclear which byte gets returned [or an error])

Yes, this is how it's done [in this helper function](https://docs.rs/embedded-hal/0.2.2/src/embedded_hal/blocking/spi.rs.html#34-55). Also with this function you can utilize RX FIFO without even without knowledge of its existence.

> The first option gets worse with shared spi devices, since some device-specific implementation could rely on the rx fifo, which would break drivers that rely on the calls matching

What do you mean under "shared spi devices"? In Rust we try to have nothing that could be shared in unsafe way.

Reply:
>  In Rust we try to have nothing that could be shared in unsafe way.

*Memory-unsafe way. If you share a device with a mutex (for example with an interrupt), you have to take care, that the `read` & `write` calls match up.

> you can utilize RX FIFO even without knowledge of its existence.

You could. But for that use case it wouldn't make much difference.

Reply:
Yes, you have to use the device in a way that prevents simultaneous access during a single transaction.

Reply:
As the trait doesn't have a notion of transactions, this is still an issue.
For example:
```
fn main() {
  let cs1 = ...;
  let cs2 = ...;
  let spi = ...;
  f1(&mut spi, &mut cs1);
  f2(&mut spi, &mut cs2);
}

// Driver 1
fn f1(spi: &mut ..., cs: &mut ...) {
    cs.try_set_low();
    block!(spi.try_send(0b00));
    cs.try_set_high();
}

// Driver 2
fn f1(spi: &mut ..., cs: &mut ...) {
    cs.try_set_low();
    block!(spi.try_send(0b00));
   // `ret` is now dependent on the implementation
   let ret = block!(spi.try_read());
   cs.try_set_high();
}
```

The only way to prevent this is to *always* pair send & read calls, nullifying any benefit of the rx fifo.